### DISCUSSION TITLE
mac 获取分辨率代码 screenshot

### DISCUSSION THREAD

**Discussion by alluxio2022** [lisai9093/YYS#12] [Q&A] [answered]
![image](https://github.com/lisai9093/YYS/assets/107172193/fdec03b4-cbb9-4d64-a2e0-74e1e5eb61f0)
返回结果为空

**Comment by alluxio2022** [comment]
Image bytes length: 2042395
First few bytes: [137  80  78  71  13  13  10  26  13  10]
解码解不出来

**Reply by lisai9093** [reply]
我在mac下用自带终端测试没有问题。你可以试试终端看看是不是IDE导致的问题？

**Reply by alluxio2022** [reply]
![image](https://github.com/lisai9093/YYS/assets/107172193/bd24c5d8-cb5b-483d-9a26-dd253400c9e4)
用终端也是一样的

**Reply by alluxio2022** [reply]
(python310) ge@MacBook-Pro YYS % adb version
Android Debug Bridge version 1.0.41
Version 34.0.5-10900879
Installed as /usr/local/bin/adb
Running on Darwin 22.3.0 (x86_64)
(python310) ge@MacBook-Pro YYS % python -V  
Python 3.10.13
(python310) ge@MacBook-Pro YYS % pip list | grep opencv
opencv-python           3.4.17.63

是我版本的问题么

**Reply by lisai9093** [reply]
ADB的版本和我是一样的
python我用的是3.11.4版本

**Comment by alluxio2022** [comment]
adb shell screencap -p /sdcard/screenshot.png
adb pull /sdcard/screenshot.png ./  
是可以打开的

**Reply by lisai9093** [reply]
我不太想直接写入/读取图片到安卓因为这样会拖慢速度，再加上screencap指令本身就比较慢。。。

**Comment by alluxio2022** [comment, marked as answer]
```python
def screenshot(monitor):
    global adb_enable, adb_path
    if adb_enable:
        command = "adb shell \"screencap -p | busybox base64\""
        pcommand = os.popen(command)
        image_bytes = pcommand.read()
        image_bytes = base64.b64decode(image_bytes)
        pcommand.close()
        if sys.platform == 'win32':
            # only for Windows, otherwise it will be None
            image_bytes = image_bytes.replace(b'\r\n', b'\n')
        screen = cv2.imdecode(numpy.frombuffer(image_bytes, numpy.uint8), cv2.IMREAD_COLOR)
        # print('screen: ',screen)
        # print('screen size: ',screen.shape[1],screen.shape[0])
        return screen
``` 
终于搞出来了

**Reply by lisai9093** [reply]
可以的，我刚刚在windows下测试也没问题，马上更新到Github，感谢！

**Reply by lisai9093** [reply]
busybox 这个是不是需要额外在安卓上安装才可以？感觉有些设备可能会不支持

**Reply by alluxio2022** [reply]
不用安装 将屏幕截图以Base64编码的形式输出到终端

**Reply by alluxio2022** [reply]
感觉是 **subprocess.run().stdout** 编码格式 和 **cv2.imdecode()** 解码格式不同造成的

**Reply by lisai9093** [reply]
有可能，你的mac是m芯片么？我手里只有Intel芯片不知道是不是和这个有关

**Reply by alluxio2022** [reply]
我也是Intel
